sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp

```diff
@@ -63,10 +63,25 @@ template <typename Compare = std::less<>> class default_sorter {
   void operator()([[maybe_unused]] Group g, [[maybe_unused]] Ptr first,
                   [[maybe_unused]] Ptr last) {
 #ifdef __SYCL_DEVICE_ONLY__
+    // Adjust the scratch pointer based on alignment of the type T.
     // Per extension specification if scratch size is less than the value
-    // returned by memory_required then behavior is undefined, so we don't check
-    // that the scratch size statisfies the requirement.
-    sycl::detail::merge_sort(g, first, last - first, comp, scratch.data());
+    // returned by memory_required, behavior is undefined. So, if std::align
+    // returns nullptr (which means that we don't have enough space) then exit
+    // early.
+    using T = typename sycl::detail::GetValueType<Ptr>::type;
+    void *scratch_ptr = scratch.data();
+    size_t space = scratch.size();
+    size_t n = last - first;
+    if (!std::align(alignof(T), n * sizeof(T), scratch_ptr, space))
+      return;
+
+    std::size_t local_id = g.get_local_linear_id();
+    T *scratch_begin =
+        local_id == 0 ? scratch_begin = ::new (scratch_ptr) T[n] : nullptr;
+    // Broadcast leader's pointer (the beginning of the scratch) to all work
+    // items in the group.
+    scratch_begin = sycl::group_broadcast(g, scratch_begin);
```

Comment:
If we're broadcasting, should we guard all the code above with `g.is_leader()` as well?

Reply:
Yes, makes sense, fixed.